File: master-server/raft/raft_log_unstable.go
package raft

import (
	"fmt"

	"master-server/raft/logger"
	"master-server/raft/proto"
)

// unstable temporary deposit the unpersistent log entries.It has log position i+unstable.offset.
// unstable can support group commit.
// Note that unstable.offset may be less than the highest log position in storage;
// this means that the next write to storage might need to truncate the log before persisting unstable.entries.
type unstable struct {
	offset uint64
	// all entries that have not yet been written to storage.
	entries []*proto.Entry
}

// maybeLastIndex returns the last index if it has at least one unstable entry.
func (u *unstable) maybeLastIndex() (uint64, bool) {
	if l := len(u.entries); l != 0 {
		return u.offset + uint64(l) - 1, true
	}
	return 0, false
}

// myabeTerm returns the term of the entry at index i, if there is any.
func (u *unstable) maybeTerm(i uint64) (uint64, bool) {
	if i < u.offset {
		return 0, false
	}

	last, ok := u.maybeLastIndex()
	if !ok || i > last {
		return 0, false
	}
	return u.entries[i-u.offset].Term, true
}

func (u *unstable) stableTo(i, t uint64) {
	gt, ok := u.maybeTerm(i)
	if !ok {
		return
	}
	if gt == t && i >= u.offset {
		l := uint64(len(u.entries))
		diff := l - (i + 1 - u.offset)
		if diff > 0 {
			copy(u.entries, u.entries[i+1-u.offset:l])
		}
		for k := diff; k < l; k++ {
			u.entries[k] = nil
		}
		u.entries = u.entries[0:diff]
		u.offset = i + 1
	}
}

func (u *unstable) restore(index uint64) {
	for i, l := 0, len(u.entries); i < l; i++ {
		u.entries[i] = nil
	}
	u.entries = u.entries[0:0]
	u.offset = index + 1
}

func (u *unstable) truncateAndAppend(ents []*proto.Entry) {
	after := ents[0].Index
	switch {
	case after == u.offset+uint64(len(u.entries)):
		// after is the next index in the u.entries directly append
		u.entries = append(u.entries, ents...)

	case after <= u.offset:
		// The log is being truncated to before our current offset portion, so set the offset and replace the entries
		for i, l := 0, len(u.entries); i < l; i++ {
			u.entries[i] = nil
		}
		u.entries = append(u.entries[0:0], ents...)
		u.offset = after

	default:
		// truncate to after and copy to u.entries then append
		u.entries = append(u.entries[0:0], u.slice(u.offset, after)...)
		u.entries = append(u.entries, ents...)
	}
}

func (u *unstable) slice(lo uint64, hi uint64) []*proto.Entry {
	u.mustCheckOutOfBounds(lo, hi)
	return u.entries[lo-u.offset : hi-u.offset]
}

// u.offset <= lo <= hi <= u.offset+len(u.offset)
func (u *unstable) mustCheckOutOfBounds(lo, hi uint64) {
	if lo > hi {
		errMsg := fmt.Sprintf("unstable.slice[%d,%d) is invalid.", lo, hi)
		logger.Error(errMsg)
		panic(AppPanicError(errMsg))
	}
	upper := u.offset + uint64(len(u.entries))
	if lo < u.offset || hi > upper {
		errMsg := fmt.Sprintf("unstable.slice[%d,%d) out of bound [%d,%d].", lo, hi, u.offset, upper)
		logger.Error(errMsg)
		panic(AppPanicError(errMsg))
	}
}

File: data-server/src/raft/src/impl/raft_log_unstable.cpp
#include "raft_log_unstable.h"

#include <sstream>
#include "raft_exception.h"

namespace sharkstore {
namespace raft {
namespace impl {

UnstableLog::UnstableLog(uint64_t offset) : offset_(offset) {}

UnstableLog::~UnstableLog() {}

bool UnstableLog::maybeLastIndex(uint64_t* last_index) const {
    if (!entries_.empty()) {
        *last_index = offset_ + entries_.size() - 1;
        return true;
    }
    return false;
}

bool UnstableLog::maybeTerm(uint64_t index, uint64_t* term) const {
    if (index < offset_) {
        return false;
    }

    uint64_t last_index = 0;
    if (maybeLastIndex(&last_index) && index <= last_index) {
        assert(index - offset_ < entries_.size() && entries_[index - offset_]);
        *term = entries_[index - offset_]->term();
        return true;
    }
    return false;
}

void UnstableLog::stableTo(uint64_t index, uint64_t term) {
    uint64_t gt = 0;
    // 确保在范围内以及term匹配
    if (!maybeTerm(index, &gt)) {
        return;
    }

    if (gt == term && index >= offset_) {
        entries_.erase(entries_.begin(), entries_.begin() + (index - offset_ + 1));
        offset_ = index + 1;
    }
}

void UnstableLog::restore(uint64_t index) {
    entries_.clear();
    offset_ = index + 1;
}

void UnstableLog::truncateAndAppend(const std::vector<EntryPtr>& ents) {
    uint64_t after = ents[0]->index();
    if (after == offset_ + static_cast<uint64_t>(entries_.size())) {
        // 直接拼接
        std::copy(ents.begin(), ents.end(), std::back_inserter(entries_));
    } else if (after <= offset_) {
        // 全部冲突，清空
        entries_.clear();
        std::copy(ents.begin(), ents.end(), std::back_inserter(entries_));
        offset_ = after;
    } else {
        // 部分冲突，截断到冲突位置
        while (!entries_.empty() && entries_.back()->index() >= after) {
            entries_.pop_back();
        }
        std::copy(ents.begin(), ents.end(), std::back_inserter(entries_));
    }
}

void UnstableLog::slice(uint64_t lo, uint64_t hi, std::vector<EntryPtr>* ents) const {
    mustCheckOutOfBounds(lo, hi);
    std::copy(entries_.begin() + (lo - offset_), entries_.begin() + (hi - offset_),
              std::back_inserter(*ents));
}

void UnstableLog::entries(std::vector<EntryPtr>* ents) const {
    std::copy(entries_.begin(), entries_.end(), std::back_inserter(*ents));
}

void UnstableLog::mustCheckOutOfBounds(uint64_t lo, uint64_t hi) const {
    if (lo > hi) {
        std::ostringstream ss;
        ss << "invalid unstable.slice:"
           << "(" << lo << "," << hi << ")";
        throw RaftException(ss.str());
    }
    auto upper = offset_ + static_cast<uint64_t>(entries_.size());
    if (lo < offset_ || hi > upper) {
        std::ostringstream ss;
        ss << "unstable.slice:"
           << "(" << lo << "," << hi << ") out of bound [" << offset_ << "," << upper
           << ",]";
        throw RaftException(ss.str());
    }
}

} /* namespace impl */
} /* namespace raft */
} /* namespace sharkstore */
